alacritty_terminal/src/term/mod.rs

```diff
@@ -1930,6 +1930,83 @@ mod tests {
 
+    #[test]
+    fn scroll_display_page_up() {
+        let size = SizeInfo::new(5., 10., 1.0, 1.0, 0.0, 0.0, false);
+        let mut term = Term::new(&MockConfig::default(), size, ());
+
+        // Create 20 lines of scrollback.
+        for _ in 0..29 {
+            term.newline();
+        }
+
+        term.scroll_display(Scroll::PageUp);
+        assert_eq!(term.vi_mode_cursor.point, Point::new(Line(-1), Column(0)));
+        assert_eq!(term.grid.display_offset(), 10);
+
+        term.scroll_display(Scroll::PageUp);
+        assert_eq!(term.vi_mode_cursor.point, Point::new(Line(-11), Column(0)));
+        assert_eq!(term.grid.display_offset(), 20);
+
+        term.scroll_display(Scroll::PageUp);
+        assert_eq!(term.vi_mode_cursor.point, Point::new(Line(-11), Column(0)));
+        assert_eq!(term.grid.display_offset(), 20);
+    }
+
+    #[test]
+    fn scroll_display_page_down() {
+        let size = SizeInfo::new(5., 10., 1.0, 1.0, 0.0, 0.0, false);
+        let mut term = Term::new(&MockConfig::default(), size, ());
+
+        // Create 20 lines of scrollback.
+        for _ in 0..29 {
+            term.newline();
+        }
+
+        // Change display_offset to topmost.
+        term.grid_mut().scroll_display(Scroll::Top);
+        term.vi_mode_cursor = ViModeCursor::new(Point::new(Line(-20), Column(0)));
+
+        term.scroll_display(Scroll::PageDown);
+        assert_eq!(term.vi_mode_cursor.point, Point::new(Line(-10), Column(0)));
+        assert_eq!(term.grid.display_offset(), 10);
+
+        term.scroll_display(Scroll::PageDown);
+        assert_eq!(term.vi_mode_cursor.point, Point::new(Line(0), Column(0)));
+        assert_eq!(term.grid.display_offset(), 0);
+
+        term.scroll_display(Scroll::PageDown);
+        assert_eq!(term.vi_mode_cursor.point, Point::new(Line(0), Column(0)));
+        assert_eq!(term.grid.display_offset(), 0);
+    }
+
+    #[test]
+    fn selection_with_scroll() {
+        let size = SizeInfo::new(5., 3., 1.0, 1.0, 0.0, 0.0, false);
+        let mut term = Term::new(&MockConfig::default(), size, ());
+
+        for _ in 0..2 {
+            term.grid_mut()[Line(2)][Column(0)].c = 'a';
+            term.grid_mut()[Line(2)][Column(1)].c = 'b';
+            term.grid_mut()[Line(2)][Column(2)].c = 'c';
+            term.clear_screen(ansi::ClearMode::All);
+        }
+        term.grid_mut()[Line(0)][Column(0)].c = 'd';
+
+        term.toggle_vi_mode();
+        term.selection =
+            Some(Selection::new(SelectionType::Lines, Point::new(Line(0), Column(0)), Side::Left));
+
+        term.scroll_display(Scroll::PageUp);
+        assert_eq!(term.selection_to_string(), Some(String::from("abc\nd\n")));
+
+        term.scroll_display(Scroll::PageUp);
+        assert_eq!(term.selection_to_string(), Some(String::from("abc\n\n\nabc\nd\n")));
+
+        term.scroll_display(Scroll::PageUp);
+        assert_eq!(term.selection_to_string(), Some(String::from("abc\n\n\nabc\nd\n")));
+    }
```

Comment:
What was the reason for adding a selection test here?

I think testing the selection logic **and** vi cursor motion logic together is usually a bit strange. I'd rather have it broken down into smaller, simpler tests which should exist already. So I'm not sure why we'd test the conversion of the selection range to string in here.

Reply:
The reason why is that `term.vi_mode_recompute_selection` is called in `term.scroll_display`. So I thought we should test whether correct re selection is completed or not. But I think your review is correct and will delete this test.